### PR TITLE
Refactor scanning of metadata and support for adding internet radio streams

### PR DESCRIPTION
**Pull request by chme**
There are two parts in this pr. The first refactors the scanning of metadata. In the current code the library knows how to scan metadata and has an ugly dependency to the filescanner. With these changes the library source is the only one who knows how to read metadata for a supported path. The library only does the general source agnostic metadata fixup.

The second part introduces a new scan_media function to the library, that takes a path that does not need to be in the library. For now only http-urls are supported. Together with the changes in mpd.c, forked-daapd is now able to add arbitrary http-streams to the queue.

For example the following mpc command now works:

```
mpc add http://path-to-radiostream
```

### CODE REVIEW

**Comment by chme** [comment]
Added a commit that fixes scanning playlists (missing http-streams in playlists). I may have gone to far with the changes in filescanner_playlist.c. This pr still needs more testing but from my side it is good for a first review.

**Comment by ejurgensen** [comment]
Ok, I will take a look at it. Sounds good that you are testing some more, just let me know when you think it is ready for merging. It would also be great if you could elaborate on the bug you found in the playlist scanner? Don't think I understood it. I also didn't quite understand what you meant about going to far with filescanner_playlist.c?

**Comment by chme** [comment]
The bug is not present in master, it is just a mistake i made during refactoring. The problem was, that i added an early continue in the loop if an http stream was already in the db and did not require a rescan. In that case the http stream was not added as a playlist item and therefor was missing in the radio section.

> I also didn't quite understand what you meant about going to far with filescanner_playlist.c?

I only meant, that i did more changes to filescanner_playlist then necessary. But i found it hard to follow the code.

File: src/library/filescanner.c [hunk 1]
```diff
@@ -539,13 +605,58 @@ create_virtual_path(char *path, char *virtual_path, int virtual_path_len)
...
+ * pointed to by 'sb'.
+ *
+ * If path is a symbolic link, the attributes in sb describe the file that the link points to
+ * and resolved_path contains the resolved path (resolved_path must be of length PATH_MAX).
+ * If path is not a symbolic link, resolved_path holds the same value as path.
+ *
+ * The return value is 0 if the operation is successful, or -1 on failure. In addition
+ */
+static int
+read_attributes(const char *path, struct stat *sb, char *resolved_path)
```
Comment (ejurgensen) [inline src/library/filescanner.c]: I'm trying to go towards having output vars as the first arguments, and then input vars. So switch path and resolved_path.

File: src/library/filescanner_playlist.c [hunk 2]
```diff
@@ -33,6 +33,7 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <errno.h>
+#include <libgen.h>
```
Comment (ejurgensen) [inline src/library/filescanner_playlist.c]: Don't see basename or dirname used here?

File: src/library/filescanner_playlist.c [hunk 3]
```diff
@@ -74,6 +75,112 @@ extinf_get(char *string, struct media_file_info *mfi, int *extinf)
...
+
+  snprintf(virtual_path, PATH_MAX, "/http:/%s", mfi->title);
+  mfi->virtual_path = strdup(virtual_path);
+
+  library_add_media(mfi);
+
+  return 0;
+}
+
+int
```
Comment (ejurgensen) [inline src/library/filescanner_playlist.c]: Should be static?

File: src/library/filescanner_playlist.c [hunk 4]
```diff
@@ -74,6 +75,112 @@ extinf_get(char *string, struct media_file_info *mfi, int *extinf)
...
+process_regular_file(char *path, char **filename)
```
Comment (ejurgensen) [inline src/library/filescanner_playlist.c]: Argument order: output first

File: src/library/filescanner.c [hunk 5]
```diff
@@ -389,37 +409,83 @@ process_deferred_playlists(void)
...
+    }
+  else
+    {
+      mfi.data_kind = DATA_KIND_FILE;
 
-	if (F_SCAN_TYPE_PIPE & type)
-	  data_kind = DATA_KIND_PIPE;
-	else
-	  data_kind = DATA_KIND_FILE;
+      if (type & F_SCAN_TYPE_AUDIOBOOK)
```
Comment (chme) [inline src/library/filescanner.c]: If it was the intention that an audiobook can also be a podcast, this woud have never worked. The media_kind is assigned either audiobook or podcast, there is and as far as i have followed the history never were a bit wise or-assignement to media_kind. The change to "else if" only makes it a bit clearer that the audiobook-flag takes precedence over the podcast-flag.

File: src/library/filescanner.c (hunk 1 above)
Comment (chme) [inline src/library/filescanner.c]: I have no real opinion on this matter, just stumbled across the guidelines from google for c/c++ code which define it the other way around: https://google.github.io/styleguide/cppguide.html#Function_Parameter_Ordering

File: src/library/filescanner.c [hunk 6]
```diff
@@ -389,37 +409,83 @@ process_deferred_playlists(void)
...
+	{
+	  free_mfi(&mfi, 1);
+	  return;
+	}
+    }
+
+  library_add_media(&mfi);
+
+  cache_artwork_ping(file, sb->st_mtime, !is_bulkscan);
+  // TODO [artworkcache] If entry in artwork cache exists for no artwork available, delete the entry if media file has embedded artwork
```
Comment (chme) [inline src/library/filescanner.c]: Yes, that is correct.

**Comment by chme** [comment]
Thanks for taking the time to review. I think, i addressed all your comments. I will do some more tests tomorrow and let you know when i am done.

**Comment by chme** [comment]
I did not notice any regressions in my tests. I think this is now ready to be merged.

**Comment by ejurgensen** [comment]
Thanks, merged. Remember to look at the TODO before you forget it.

It's true that forked-daapd doesn't do bitwise media_kind's, and even if it did. I'm not sure it would work with the daap queries. But since Apple did select powers of 2 I suppose their original intention was to allow combinations.
